Userland/userdel.cpp

```diff
@@ -0,0 +1,229 @@
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * 
+ * 
+ * The code for function remove() was taken from rm.cpp in this directory.
+ * The copyright notice is as follows:
+ *
+ * Copyright (c) 2018-2020, Andreas Kling <kling@serenityos.org>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <AK/String.h>
+#include <AK/StringBuilder.h>
+#include <LibCore/ArgsParser.h>
+#include <ctype.h>
+#include <dirent.h>
+#include <errno.h>
+#include <pwd.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+static void parseArg(int argc, char** argv, const char*& username, bool& removeHome);
+static int deleteUser(const char* username, bool removeHome);
+static int remove(bool recursive, String path);
+
+int main(int argc, char** argv)
+{
+    const char* username = nullptr;
+    bool removeHome = false;
+    parseArg(argc, argv, username, removeHome);
+
+    return deleteUser(username, removeHome);
+}
+
+void parseArg(int argc, char** argv, const char*& username, bool& removeHome)
+{
+    Core::ArgsParser args_parser;
+    args_parser.add_option(removeHome, "Remove home directory", "remove", 'r');
+    args_parser.add_positional_argument(username, "Login user identity (username)", "login");
+    args_parser.parse(argc, argv);
+}
+
+int deleteUser(const char* username, bool removeHome)
+{
+    FILE* pwfile = fopen("/etc/passwd", "r+");
```

Comment:
Instead of opening `/etc/passwd` and manipulating it manually, let's instead use the standard C library functions:

* `setpwent()`
* `getpwent()`
* `putpwent()`

To remove a given username, you could leverage existing functionality like so:

```cpp
char path[] = "/etc/passwd.XXXXXX";
auto fd = mkstemp(path);
auto* new_passwd = fdopen(fd, "w");
setpwent();
for (auto* pw = getpwent(); pw; pw = getpwent()) {
    if (!strcmp(pw->pw_name, user_to_remove))
        continue;
    if (putpwent(pw, new_passwd) < 0) {
        // handle error
    }
}
if (rename(path, "/etc/passwd") < 0) {
    // handle error
}
fclose(new_passwd);
```

For removing the files, I would suggest invoking `/bin/rm` rather than implementing removal manually.
That way we don't need to maintain two copies of that code. :)